File: articles/nonlinearity-feedback-biological-regulation/python/delayed_feedback.py
"""
Delayed negative feedback scenarios.

Run:
    python python/delayed_feedback.py
"""

from pathlib import Path

import pandas as pd

from nonlinear_feedback_core import simulate_delayed_negative_feedback


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "delayed_feedback_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        trajectory = simulate_delayed_negative_feedback(
            x0=row["x0"],
            production_rate=row["production_rate"],
            feedback_strength=row["feedback_strength"],
            delay=row["delay"],
            dt=row["dt"],
            t_end=row["t_end"],
        )

        rows.append(
            {
                "scenario": row["scenario"],
                "delay": row["delay"],
                "final_state": trajectory["state"].iloc[-1],
                "max_state": trajectory["state"].max(),
                "state_range": trajectory["state"].max() - trajectory["state"].min(),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/nonlinearity-feedback-biological-regulation/r/delayed_feedback.R
# Delayed negative feedback scaffold.

simulate_delayed_negative_feedback <- function(x0, production_rate, feedback_strength, delay, dt = 0.01, t_end = 80) {
  time <- seq(0, t_end, by = dt)
  x <- numeric(length(time))
  x[1] <- x0
  delay_steps <- max(as.integer(delay / dt), 1)

  for (i in 2:length(time)) {
    delayed_index <- max(i - delay_steps, 1)
    delayed_state <- x[delayed_index]
    dx <- production_rate - feedback_strength * delayed_state
    x[i] <- max(x[i - 1] + dx * dt, 0)
  }

  data.frame(time = time, state = x)
}

delays <- c(0.1, 1.0, 4.0, 8.0)

rows <- lapply(delays, function(delay) {
  trajectory <- simulate_delayed_negative_feedback(1.0, 1.0, 0.8, delay)

  data.frame(
    delay = delay,
    final_state = tail(trajectory$state, 1),
    max_state = max(trajectory$state),
    state_range = max(trajectory$state) - min(trajectory$state)
  )
})

summary_df <- do.call(rbind, rows)

print(round(summary_df, 5))
